<a href="https://colab.research.google.com/github/JewareSM/Data-Visualization-Tools/blob/main/DSML_Advanced_Pandas_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd

In [ ]:
#Business case study for today -

''' 
    Imagine that you are working as a DS for Mickinsy Digital.
    Mickinsy Digital wants to understand relationship between GDP per capita and life expectancy.
    Life expectancy - Means average age of a person lives in that coutry would be like life expectancy.
    Average age for the population would be life expectancy.

'''

In [ ]:
#Let's start with reading a dataset -

!gdown 1E3bwvYGf1ig32RmcYiWc0IXPN-mD_bI_

Downloading...
From: https://drive.google.com/uc?id=1E3bwvYGf1ig32RmcYiWc0IXPN-mD_bI_
To: /content/gapminder.csv
100% 83.8k/83.8k [00:00<00:00, 69.3MB/s]


* Dataset Link - https://drive.google.com/file/d/1E3bwvYGf1ig32RmcYiWc0IXPN-mD_bI_/view?usp=sharing

In [ ]:
#Let's read dataset -

pd.read_csv("gapminder.csv")           #Pandas provide read function to read data set. ---->> pd.read_(type of file like .txt file or .csv file)(file_name)

country  year  population continent  life_exp     gdp_cap
0     Afghanistan  1952     8425333      Asia    28.801  779.445314
1     Afghanistan  1957     9240934      Asia    30.332  820.853030
2     Afghanistan  1962    10267083      Asia    31.997  853.100710
3     Afghanistan  1967    11537966      Asia    34.020  836.197138
4     Afghanistan  1972    13079460      Asia    36.088  739.981106
...           ...   ...         ...       ...       ...         ...
1699     Zimbabwe  1987     9216418    Africa    62.351  706.157306
1700     Zimbabwe  1992    10704340    Africa    60.377  693.420786
1701     Zimbabwe  1997    11404948    Africa    46.809  792.449960
1702     Zimbabwe  2002    11926563    Africa    39.989  672.038623
1703     Zimbabwe  2007    12311143    Africa    43.487  469.709298

[1704 rows x 6 columns]

* This data is in CSV format and file name is "gapminder.csv" , it's a public dataset avaliable .

* This data set has 1704 rows and 7 columns.

In [ ]:
#Let's me save this data in varible 'df'

df = pd.read_csv("gapminder.csv")

In [ ]:
df             #Structured data or Tabular data.

country  year  population continent  life_exp     gdp_cap
0     Afghanistan  1952     8425333      Asia    28.801  779.445314
1     Afghanistan  1957     9240934      Asia    30.332  820.853030
2     Afghanistan  1962    10267083      Asia    31.997  853.100710
3     Afghanistan  1967    11537966      Asia    34.020  836.197138
4     Afghanistan  1972    13079460      Asia    36.088  739.981106
...           ...   ...         ...       ...       ...         ...
1699     Zimbabwe  1987     9216418    Africa    62.351  706.157306
1700     Zimbabwe  1992    10704340    Africa    60.377  693.420786
1701     Zimbabwe  1997    11404948    Africa    46.809  792.449960
1702     Zimbabwe  2002    11926563    Africa    39.989  672.038623
1703     Zimbabwe  2007    12311143    Africa    43.487  469.709298

[1704 rows x 6 columns]

* ' df '  is a short form for dataframe.

* Datafeame - Its a data structure provided by pandas which can help you stroing a structured data or tabular data.

In [ ]:
#Let's check type of object 'df'

type(df)

pandas.core.frame.DataFrame

* So type of this object is DataFrame.

* What is pandas dataframe - Pandas dataframe is basically a table like representation of data.
* It's a data structure provided by pandas which helps us to store a data in Structured format or Tabular format.

* You can consider it as counter part of a matrices in numpy.
* Why matrices in numpy ?? --->> Because it has 2 dimensions , First dimension is along 0th dimension , which is rows(Axis = 0) , 
Another dimension lets say 'index 1' , which is columns (Axis = 1) .
So it is 2D object so we can think it as a matrix counter part of numpy.

In [ ]:
#Let's say if you want to access one of the colume from pandas dataframe -

df["country"]

0       Afghanistan
1       Afghanistan
2       Afghanistan
3       Afghanistan
4       Afghanistan
           ...     
1699       Zimbabwe
1700       Zimbabwe
1701       Zimbabwe
1702       Zimbabwe
1703       Zimbabwe
Name: country, Length: 1704, dtype: object

* You have gotten one of the column.

* And see that name of column , length of column and dtype of column is mentioned.

* dtype of column is 'object' we will know what object it is.

* Last think to be noted is that we have also labels or indexes are also provided along with every value in the column.

In [ ]:
#Let's check type of 'country' column

type(df["country"])

pandas.core.series.Series

* It tells us a type of column is a ' pandas series ' .

* What pandas series is?? ---->> Pandas series is another data structure provided by pandas libabry .

* You can think 'pandas series' as a vector.

* Note -
* 1. You can think 'dataframe' as a Matrix .(Numpy_array)
* 2. You can think 'pandas series'as a vector . (Numpy_vector)

* As we learn ' dataframe ' is 2D object and ' pandas series 'is 1D object .  So , here we are providing some analogy for better understanding , there is no relation between them .


* A one single column is a ' pands series ' and if you stack multiple series together you will get ' pandas dataframe ' .

* If we look at this data again , I see couple of rows are missing overhere in this display. Of course this data has clse to 1700 rows and if we start printing all 1700 rows this would take a lot of displaying area.

* So it's not possiable to see all your data in one go , using this( Google colab or Jupyter notebook ) kind of display .

* So what we can do to get summary of data , 
So we can use method called as **info ()** .


* This info () function gives information about columns if they have some missing values and there data types.

In [ ]:
#Let's check missing values in data and data types of all columns -

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1704 entries, 0 to 1703
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   country     1704 non-null   object 
 1   year        1704 non-null   int64  
 2   population  1704 non-null   int64  
 3   continent   1704 non-null   object 
 4   life_exp    1704 non-null   float64
 5   gdp_cap     1704 non-null   float64
dtypes: float64(2), int64(2), object(2)
memory usage: 80.0+ KB


 * From above output we can see there are 6 columns , labled from 0 to 5 .

 * It also tells us non - null count . (Non - missing values count)
 * It is basically count of non missing values and if you notice this , this is same number of entries in data . You can see all the non - null values in data are equal to 1704 .

* We can write a fancy code to check missing values in dataset , we will learn this in future . But you can check this using .info () function .

* Last column in above output is ' Dtype '.
* When you see dtype as a ' object ' you can corresponds it to ' strings '  in most of the cases . 

In [ ]:
#Let's print top row of dataset

In [ ]:
df.head()

country  year  population continent  life_exp     gdp_cap
0  Afghanistan  1952     8425333      Asia    28.801  779.445314
1  Afghanistan  1957     9240934      Asia    30.332  820.853030
2  Afghanistan  1962    10267083      Asia    31.997  853.100710
3  Afghanistan  1967    11537966      Asia    34.020  836.197138
4  Afghanistan  1972    13079460      Asia    36.088  739.981106

* By default this will print top five rows for us .
* We can customise object behaviour .


In [ ]:
#Let's print top 11 rows

df.head(11)

country  year  population continent  life_exp     gdp_cap
0   Afghanistan  1952     8425333      Asia    28.801  779.445314
1   Afghanistan  1957     9240934      Asia    30.332  820.853030
2   Afghanistan  1962    10267083      Asia    31.997  853.100710
3   Afghanistan  1967    11537966      Asia    34.020  836.197138
4   Afghanistan  1972    13079460      Asia    36.088  739.981106
5   Afghanistan  1977    14880372      Asia    38.438  786.113360
6   Afghanistan  1982    12881816      Asia    39.854  978.011439
7   Afghanistan  1987    13867957      Asia    40.822  852.395945
8   Afghanistan  1992    16317921      Asia    41.674  649.341395
9   Afghanistan  1997    22227415      Asia    41.763  635.341351
10  Afghanistan  2002    25268405      Asia    42.129  726.734055

In [ ]:
#Let's print bottem rows of dataset

df.tail()               #By default this will print bottem five rows for us .

country  year  population continent  life_exp     gdp_cap
1699  Zimbabwe  1987     9216418    Africa    62.351  706.157306
1700  Zimbabwe  1992    10704340    Africa    60.377  693.420786
1701  Zimbabwe  1997    11404948    Africa    46.809  792.449960
1702  Zimbabwe  2002    11926563    Africa    39.989  672.038623
1703  Zimbabwe  2007    12311143    Africa    43.487  469.709298

In [ ]:
#Let's print bottem 7 rows 

df.tail(7)

country  year  population continent  life_exp     gdp_cap
1697  Zimbabwe  1977     6642107    Africa    57.674  685.587682
1698  Zimbabwe  1982     7636524    Africa    60.363  788.855041
1699  Zimbabwe  1987     9216418    Africa    62.351  706.157306
1700  Zimbabwe  1992    10704340    Africa    60.377  693.420786
1701  Zimbabwe  1997    11404948    Africa    46.809  792.449960
1702  Zimbabwe  2002    11926563    Africa    39.989  672.038623
1703  Zimbabwe  2007    12311143    Africa    43.487  469.709298

In [ ]:
#We can check shape of dataset --

df.shape

(1704, 6)

* We have 1704 entries along 1st dimension ( Axis = 0) which is rows and 6 entries along 2nd dimension ( Axis = 1 ) which is columns .

In [ ]:
#Let's try to get more detailed summary of data for that we can use 'df.describe()' function 

df.describe()

year    population     life_exp        gdp_cap
count  1704.00000  1.704000e+03  1704.000000    1704.000000
mean   1979.50000  2.960121e+07    59.474439    7215.327081
std      17.26533  1.061579e+08    12.917107    9857.454543
min    1952.00000  6.001100e+04    23.599000     241.165876
25%    1965.75000  2.793664e+06    48.198000    1202.060309
50%    1979.50000  7.023596e+06    60.712500    3531.846988
75%    1993.25000  1.958522e+07    70.845500    9325.462346
max    2007.00000  1.318683e+09    82.603000  113523.132900

* We can see that this summary is provided for numerical data only , because calculating this metrice for ' categorical ' data or for ' object ' data type is not possiable .


* From above output we say the average life expectancy of all countries is 59.474439 .

* Standard deviation is close to 13 which means there is a lot of variation across different countries .

* The average GDP_cap for a country is 7215.327081 $$ .

* Median GDP_cap is 3531.846988 $$ .

In [ ]:
#Let's check more detailed summary of other features whose data type is object .
''' For that we modify describe() function '''

df.describe(include = "object")

country continent
count          1704      1704
unique          142         5
top     Afghanistan    Africa
freq             12       624

* You can modify describe () function , descibe function has argument called as ' include ' .
* In this , ' include ' argument you can pass string or list of string to get description / summary  of features whose data type is ' object ' .

* From above output you can we get summary of country and continent columns whose dtype is object .


* From above output we can say - There are 142 unique countries are present in our data set and 5 unique continent .

* Most frequently occuring country is ' Afghanistan ' and most occuring continent is ' Africa '  .

* And there frequency count is also present - 12 times Afghanistan occures ans 624 times Africa occures .


In [ ]:
#Basics operatoins on columns 

''' We have our dataset reffred by 'df' variable , so how can we extract column names using this variable '''

df.columns


Index(['country', 'year', 'population', 'continent', 'life_exp', 'gdp_cap'], dtype='object')

In [ ]:
df.keys()          #Key() function

Index(['country', 'year', 'population', 'continent', 'life_exp', 'gdp_cap'], dtype='object')

In [ ]:
#Before this we have used concept of keys in dictionary

df["country"]

0       Afghanistan
1       Afghanistan
2       Afghanistan
3       Afghanistan
4       Afghanistan
           ...     
1699       Zimbabwe
1700       Zimbabwe
1701       Zimbabwe
1702       Zimbabwe
1703       Zimbabwe
Name: country, Length: 1704, dtype: object

* Pandas dataframe is known as specialied dictionary .

* Why pandas dataframe is known as specialied dictionary ?? --- >> Because in pandas dictionary you can provide multiple keys .


* In normal dictionary it is not possiable .

In [ ]:
df[["country", "life_exp"]]        #Specialied dictionary

country  life_exp
0     Afghanistan    28.801
1     Afghanistan    30.332
2     Afghanistan    31.997
3     Afghanistan    34.020
4     Afghanistan    36.088
...           ...       ...
1699     Zimbabwe    62.351
1700     Zimbabwe    60.377
1701     Zimbabwe    46.809
1702     Zimbabwe    39.989
1703     Zimbabwe    43.487

[1704 rows x 2 columns]

* So  here we multiple data in a one single line of code .

In [ ]:
#If you want find all of the countries , which have been surveyed 

''' Every  country should only appear once so you can use unique() method on country column '''

df["country"].unique()

array(['Afghanistan', 'Albania', 'Algeria', 'Angola', 'Argentina',
       'Australia', 'Austria', 'Bahrain', 'Bangladesh', 'Belgium',
       'Benin', 'Bolivia', 'Bosnia and Herzegovina', 'Botswana', 'Brazil',
       'Bulgaria', 'Burkina Faso', 'Burundi', 'Cambodia', 'Cameroon',
       'Canada', 'Central African Republic', 'Chad', 'Chile', 'China',
       'Colombia', 'Comoros', 'Congo, Dem. Rep.', 'Congo, Rep.',
       'Costa Rica', "Cote d'Ivoire", 'Croatia', 'Cuba', 'Czech Republic',
       'Denmark', 'Djibouti', 'Dominican Republic', 'Ecuador', 'Egypt',
       'El Salvador', 'Equatorial Guinea', 'Eritrea', 'Ethiopia',
       'Finland', 'France', 'Gabon', 'Gambia', 'Germany', 'Ghana',
       'Greece', 'Guatemala', 'Guinea', 'Guinea-Bissau', 'Haiti',
       'Honduras', 'Hong Kong, China', 'Hungary', 'Iceland', 'India',
       'Indonesia', 'Iran', 'Iraq', 'Ireland', 'Israel', 'Italy',
       'Jamaica', 'Japan', 'Jordan', 'Kenya', 'Korea, Dem. Rep.',
       'Korea, Rep.', 'Kuwait', 'Leba

In [ ]:
#If we want to check count of unique countries 

len(df["country"].unique())

142

* So there are 142 unique country present in our dataset .

In [ ]:
''' We might also be inserested to know how many times let's say 'Afghanistan' is appearing in our data.
    Let's say how many times 'Urgentina' is appearing in our data .
    Basically we are not looking only for the unique value in data, we are also looking for there connts '''


df["country"].value_counts()

Afghanistan          12
Pakistan             12
New Zealand          12
Nicaragua            12
Niger                12
                     ..
Eritrea              12
Equatorial Guinea    12
El Salvador          12
Egypt                12
Zimbabwe             12
Name: country, Length: 142, dtype: int64

* Here we get list of countries and there respective counts .

In [ ]:
#Let's check type of this 

type(df["country"].value_counts())

pandas.core.series.Series

* In df["country"].value_counts( ) we have countries and there respective counts , but type of this is 'series'.

* It's a series then it supposed to be a 1D array of object .

* So here data stored in series is actually 'value_counts' and what you see in left (country column) are just indexes .

* So series can have indexes  which can be non numberes also .

In [ ]:
#Let's check indexes now

df["country"].value_counts().index

Index(['Afghanistan', 'Pakistan', 'New Zealand', 'Nicaragua', 'Niger',
       'Nigeria', 'Norway', 'Oman', 'Panama', 'Nepal',
       ...
       'Gambia', 'Gabon', 'France', 'Finland', 'Ethiopia', 'Eritrea',
       'Equatorial Guinea', 'El Salvador', 'Egypt', 'Zimbabwe'],
      dtype='object', length=142)

In [ ]:
#Let's we want to rename a particular column

''' So for this you can use rename() function '''

df.rename({"country":"Country"},axis = 1)

Country  year  population continent  life_exp     gdp_cap
0     Afghanistan  1952     8425333      Asia    28.801  779.445314
1     Afghanistan  1957     9240934      Asia    30.332  820.853030
2     Afghanistan  1962    10267083      Asia    31.997  853.100710
3     Afghanistan  1967    11537966      Asia    34.020  836.197138
4     Afghanistan  1972    13079460      Asia    36.088  739.981106
...           ...   ...         ...       ...       ...         ...
1699     Zimbabwe  1987     9216418    Africa    62.351  706.157306
1700     Zimbabwe  1992    10704340    Africa    60.377  693.420786
1701     Zimbabwe  1997    11404948    Africa    46.809  792.449960
1702     Zimbabwe  2002    11926563    Africa    39.989  672.038623
1703     Zimbabwe  2007    12311143    Africa    43.487  469.709298

[1704 rows x 6 columns]

* We can see 'country' get replace to captial 'Country' .


In [ ]:
#Lets print dataset again

df

country  year  population continent  life_exp     gdp_cap
0     Afghanistan  1952     8425333      Asia    28.801  779.445314
1     Afghanistan  1957     9240934      Asia    30.332  820.853030
2     Afghanistan  1962    10267083      Asia    31.997  853.100710
3     Afghanistan  1967    11537966      Asia    34.020  836.197138
4     Afghanistan  1972    13079460      Asia    36.088  739.981106
...           ...   ...         ...       ...       ...         ...
1699     Zimbabwe  1987     9216418    Africa    62.351  706.157306
1700     Zimbabwe  1992    10704340    Africa    60.377  693.420786
1701     Zimbabwe  1997    11404948    Africa    46.809  792.449960
1702     Zimbabwe  2002    11926563    Africa    39.989  672.038623
1703     Zimbabwe  2007    12311143    Africa    43.487  469.709298

[1704 rows x 6 columns]

* Surprisingly the changes are not reflected in the dataset .

* So if you want make this changes permanently you can use "inplace" or you can pass this argument called as 'inplace' in rename() function that will make permanent changes in the dataset .

In [ ]:
df.rename({"country":"Country"},axis = 1,inplace = True)

In [ ]:
df

Country  year  population continent  life_exp     gdp_cap
0     Afghanistan  1952     8425333      Asia    28.801  779.445314
1     Afghanistan  1957     9240934      Asia    30.332  820.853030
2     Afghanistan  1962    10267083      Asia    31.997  853.100710
3     Afghanistan  1967    11537966      Asia    34.020  836.197138
4     Afghanistan  1972    13079460      Asia    36.088  739.981106
...           ...   ...         ...       ...       ...         ...
1699     Zimbabwe  1987     9216418    Africa    62.351  706.157306
1700     Zimbabwe  1992    10704340    Africa    60.377  693.420786
1701     Zimbabwe  1997    11404948    Africa    46.809  792.449960
1702     Zimbabwe  2002    11926563    Africa    39.989  672.038623
1703     Zimbabwe  2007    12311143    Africa    43.487  469.709298

[1704 rows x 6 columns]

* Now we can see 'country' get replace to captial 'Country' parmanetly .

* Note - Check discription of rename( ) function.
* In rename function you can pass multiple arguments as a dictionary . 

In [ ]:
#Lets get our original dataset 

df.rename({"Country":"country"},axis = 1,inplace = True)

In [ ]:
df

country  year  population continent  life_exp     gdp_cap
0     Afghanistan  1952     8425333      Asia    28.801  779.445314
1     Afghanistan  1957     9240934      Asia    30.332  820.853030
2     Afghanistan  1962    10267083      Asia    31.997  853.100710
3     Afghanistan  1967    11537966      Asia    34.020  836.197138
4     Afghanistan  1972    13079460      Asia    36.088  739.981106
...           ...   ...         ...       ...       ...         ...
1699     Zimbabwe  1987     9216418    Africa    62.351  706.157306
1700     Zimbabwe  1992    10704340    Africa    60.377  693.420786
1701     Zimbabwe  1997    11404948    Africa    46.809  792.449960
1702     Zimbabwe  2002    11926563    Africa    39.989  672.038623
1703     Zimbabwe  2007    12311143    Africa    43.487  469.709298

[1704 rows x 6 columns]

In [ ]:
df["country"]     #Dictionary way to access column

0       Afghanistan
1       Afghanistan
2       Afghanistan
3       Afghanistan
4       Afghanistan
           ...     
1699       Zimbabwe
1700       Zimbabwe
1701       Zimbabwe
1702       Zimbabwe
1703       Zimbabwe
Name: country, Length: 1704, dtype: object

In [ ]:
df.country        #Attribute way to access column

0       Afghanistan
1       Afghanistan
2       Afghanistan
3       Afghanistan
4       Afghanistan
           ...     
1699       Zimbabwe
1700       Zimbabwe
1701       Zimbabwe
1702       Zimbabwe
1703       Zimbabwe
Name: country, Length: 1704, dtype: object

* Attribute way to access column is not recommended to use . Because -
* Column name can have some space in between .
* What if name the of the column is same as one of the attribute then it will through an error . we will be not able to seperate which column name which is attribute . 
* If column names are contain number or any special character .

* So don't use  attribute way to access column .

In [ ]:
#How to drop/delete a particular column from dataset 

''' For this you can use drop() function.
    In drop function first you have to pass column name that you wants to drop and then axis '''

df.drop("continent",axis = 1)  

country  year  population  life_exp     gdp_cap
0     Afghanistan  1952     8425333    28.801  779.445314
1     Afghanistan  1957     9240934    30.332  820.853030
2     Afghanistan  1962    10267083    31.997  853.100710
3     Afghanistan  1967    11537966    34.020  836.197138
4     Afghanistan  1972    13079460    36.088  739.981106
...           ...   ...         ...       ...         ...
1699     Zimbabwe  1987     9216418    62.351  706.157306
1700     Zimbabwe  1992    10704340    60.377  693.420786
1701     Zimbabwe  1997    11404948    46.809  792.449960
1702     Zimbabwe  2002    11926563    39.989  672.038623
1703     Zimbabwe  2007    12311143    43.487  469.709298

[1704 rows x 5 columns]

* Continent column is droped .

In [ ]:
df

country  year  population continent  life_exp     gdp_cap
0     Afghanistan  1952     8425333      Asia    28.801  779.445314
1     Afghanistan  1957     9240934      Asia    30.332  820.853030
2     Afghanistan  1962    10267083      Asia    31.997  853.100710
3     Afghanistan  1967    11537966      Asia    34.020  836.197138
4     Afghanistan  1972    13079460      Asia    36.088  739.981106
...           ...   ...         ...       ...       ...         ...
1699     Zimbabwe  1987     9216418    Africa    62.351  706.157306
1700     Zimbabwe  1992    10704340    Africa    60.377  693.420786
1701     Zimbabwe  1997    11404948    Africa    46.809  792.449960
1702     Zimbabwe  2002    11926563    Africa    39.989  672.038623
1703     Zimbabwe  2007    12311143    Africa    43.487  469.709298

[1704 rows x 6 columns]

In [ ]:
#If you want to drop a particular column permanently then pass argument 'inplace = True' in drop function it will drop a particular column permanently 

df.drop("continent",axis = 1 ,inplace = True) //

''' Out [] : It will drop continent column permanently 

    ----* Don't run this cell *-----  '''

In [ ]:
#Let's see how we can add a new column in dataset 

''' Lets add a new column 'total_gdp' in dataset.
    You can calculate total_gdp by -
    
    total_gdp = population * gdp_cap '''


df["total_gdp"] = df["population"] * df["gdp_cap"]

In [ ]:
df

country  year  population continent  life_exp     gdp_cap  \
0     Afghanistan  1952     8425333      Asia    28.801  779.445314   
1     Afghanistan  1957     9240934      Asia    30.332  820.853030   
2     Afghanistan  1962    10267083      Asia    31.997  853.100710   
3     Afghanistan  1967    11537966      Asia    34.020  836.197138   
4     Afghanistan  1972    13079460      Asia    36.088  739.981106   
...           ...   ...         ...       ...       ...         ...   
1699     Zimbabwe  1987     9216418    Africa    62.351  706.157306   
1700     Zimbabwe  1992    10704340    Africa    60.377  693.420786   
1701     Zimbabwe  1997    11404948    Africa    46.809  792.449960   
1702     Zimbabwe  2002    11926563    Africa    39.989  672.038623   
1703     Zimbabwe  2007    12311143    Africa    43.487  469.709298   

         total_gdp  
0     6.567086e+09  
1     7.585449e+09  
2     8.758856e+09  
3     9.648014e+09  
4     9.678553e+09  
...            ...  
1699  6.508241e+09  
1700  7.422612e+09  
1701  9.037851e+09  
1702  8.015111e+09  
1703  5.782658e+09  

[1704 rows x 7 columns]

* New column total_gdp is added to dataset .

In [ ]:
''' There can be instances where I want to create a new feature but that has no relation between current features .
    i.e is something that we have create from scratch 
    
    So for that we can create list or array or series and assigne it as one column .'''


df["custome"] = (i for i in range(1704))    

TypeError: ignored

In [ ]:
df["custom"] = [i for i in range(1704)]

In [ ]:
df

country  year  population continent  life_exp     gdp_cap  \
0     Afghanistan  1952     8425333      Asia    28.801  779.445314   
1     Afghanistan  1957     9240934      Asia    30.332  820.853030   
2     Afghanistan  1962    10267083      Asia    31.997  853.100710   
3     Afghanistan  1967    11537966      Asia    34.020  836.197138   
4     Afghanistan  1972    13079460      Asia    36.088  739.981106   
...           ...   ...         ...       ...       ...         ...   
1699     Zimbabwe  1987     9216418    Africa    62.351  706.157306   
1700     Zimbabwe  1992    10704340    Africa    60.377  693.420786   
1701     Zimbabwe  1997    11404948    Africa    46.809  792.449960   
1702     Zimbabwe  2002    11926563    Africa    39.989  672.038623   
1703     Zimbabwe  2007    12311143    Africa    43.487  469.709298   

         total_gdp  custom  
0     6.567086e+09       0  
1     7.585449e+09       1  
2     8.758856e+09       2  
3     9.648014e+09       3  
4     9.678553e+09       4  
...            ...     ...  
1699  6.508241e+09    1699  
1700  7.422612e+09    1700  
1701  9.037851e+09    1701  
1702  8.015111e+09    1702  
1703  5.782658e+09    1703  

[1704 rows x 8 columns]

* This is basically a list that we have assigned to a column , similarly we can assign array or series also and we can assign it as a column .

In [ ]:
#Let's delete columns that we have created for our understanding to get our original data set back because we want work on that .

df.drop(["total_gdp","custom"],axis = 1,inplace = True)

In [ ]:
df

country  year  population continent  life_exp     gdp_cap
0     Afghanistan  1952     8425333      Asia    28.801  779.445314
1     Afghanistan  1957     9240934      Asia    30.332  820.853030
2     Afghanistan  1962    10267083      Asia    31.997  853.100710
3     Afghanistan  1967    11537966      Asia    34.020  836.197138
4     Afghanistan  1972    13079460      Asia    36.088  739.981106
...           ...   ...         ...       ...       ...         ...
1699     Zimbabwe  1987     9216418    Africa    62.351  706.157306
1700     Zimbabwe  1992    10704340    Africa    60.377  693.420786
1701     Zimbabwe  1997    11404948    Africa    46.809  792.449960
1702     Zimbabwe  2002    11926563    Africa    39.989  672.038623
1703     Zimbabwe  2007    12311143    Africa    43.487  469.709298

[1704 rows x 6 columns]

In [ ]:
#Working with rows -

In [ ]:
ser = df["country"]

In [ ]:
ser

0       Afghanistan
1       Afghanistan
2       Afghanistan
3       Afghanistan
4       Afghanistan
           ...     
1699       Zimbabwe
1700       Zimbabwe
1701       Zimbabwe
1702       Zimbabwe
1703       Zimbabwe
Name: country, Length: 1704, dtype: object

In [ ]:
#Let's say if we want to access the 6th row 

ser[6]            #This is example of indexing

'Afghanistan'

In [ ]:
#Let's access 4th row

ser[3]            #This is example of indexing

'Afghanistan'

In [ ]:
#Let's say we have to access 4th row to 15th row 

ser[4:16]          #This is example of slicing

4     Afghanistan
5     Afghanistan
6     Afghanistan
7     Afghanistan
8     Afghanistan
9     Afghanistan
10    Afghanistan
11    Afghanistan
12        Albania
13        Albania
14        Albania
15        Albania
Name: country, dtype: object

In [ ]:
#Let's assume a scenario that I'm not happy with above indexing and I want to change it to some thing else

''' Currently it is zero based indexing and I want to change it to one based indexing '''

#Let's first understand how we can extract indexes 


ser.index

RangeIndex(start=0, stop=1704, step=1)

* We are not happy with positional indexing .
* Positional indexing - positional index defines the position in that particular area .
* Example - array 
* Array is positioned these indexes 0,1,2,3..... 
* Array start there indexing from 0 onwards .
* Its giving us a generator because it happens to be a big value .

In [ ]:
#Now lets change zero based indexing to one based indexing

import numpy as np

In [ ]:
ser.index = np.arange(1, 1705, 1)

In [ ]:
ser         #ser is short form of 'Series'

1       Afghanistan
2       Afghanistan
3       Afghanistan
4       Afghanistan
5       Afghanistan
           ...     
1700       Zimbabwe
1701       Zimbabwe
1702       Zimbabwe
1703       Zimbabwe
1704       Zimbabwe
Name: country, Length: 1704, dtype: object

In [ ]:
ser.index = np.arange(1, df.shape[0]+1, 1)

In [ ]:
ser        #ser is short form of 'Series'

1       Afghanistan
2       Afghanistan
3       Afghanistan
4       Afghanistan
5       Afghanistan
           ...     
1700       Zimbabwe
1701       Zimbabwe
1702       Zimbabwe
1703       Zimbabwe
1704       Zimbabwe
Name: country, Length: 1704, dtype: object

In [ ]:
type(ser)

pandas.core.series.Series

* Again this is series .

In [ ]:
ser.index

Int64Index([   1,    2,    3,    4,    5,    6,    7,    8,    9,   10,
            ...
            1695, 1696, 1697, 1698, 1699, 1700, 1701, 1702, 1703, 1704],
           dtype='int64', length=1704)

In [ ]:
#Implicit and Explicit Indices

ser.index           #Explicit Indices or Row labels or Row indexes or May be Row only

Int64Index([   1,    2,    3,    4,    5,    6,    7,    8,    9,   10,
            ...
            1695, 1696, 1697, 1698, 1699, 1700, 1701, 1702, 1703, 1704],
           dtype='int64', length=1704)

* We have provided it ( one based indexing ) from outside so this is called as Explicit Indexes .
* Explicit Indexes can be strings or numbers .

* Explicit Indexes also called as Row labels or Row indexes or May be Row only .

* The indexes that are visible , when you print dataframe is called as Positioanl indexes .

* Positional indexes always start from zero and so on .
* Positional indexes are always integeres .
* Positional indexes is also called as Implicit indexes .

In [ ]:
ser.index         #Row labels

Int64Index([   1,    2,    3,    4,    5,    6,    7,    8,    9,   10,
            ...
            1695, 1696, 1697, 1698, 1699, 1700, 1701, 1702, 1703, 1704],
           dtype='int64', length=1704)

In [ ]:
#You can check what is the index of 2nd row

ser.index[1]

2

In [ ]:
#Let's do something more interesting about this indexes.

''' For that we are going to create a new series from scratch and we will learn things from series and then we will translate all that information to bigger dataset.'''

data = pd.Series(["a", "b", "c"], index=[1, 5, 3])

In [ ]:
data


1    a
5    b
3    c
dtype: object

In [ ]:
#Let's check type of data

type(data)

pandas.core.series.Series

* This is series and content of the series is here ( Right side part) .
* In left side we have 1 , 5 , 3 that are indices or The explicit indices or Labels .

In [ ]:
#Let's do some Indexing and Slicing over this series 

data[1]     #Indexing uses explicit indices

'a'

In [ ]:
data[1:3]   #Slicing uses implicit or positional indices

5    b
3    c
dtype: object

* Note - 

* 1. When your doing ' Indexing ' then it will use explicit indices .
* 2.When your doing ' Slicing 'then it will use implicit or positional indices .


* Now this thing is totally maddening , so that' s why we will come to a solution , we learn something ' lock ' and ' ilock ' which will helps us to not have these kind of problem .


In [ ]:
#Let's see one more example of series 


data = pd.Series(['a', 'b', 'c'], index=['x', 'y', 'z'])

In [ ]:
data

x    a
y    b
z    c
dtype: object

In [ ]:
#Let's do some Indexing and Slicing over this series 


data["x"]                    #Indexing uses explicit indexing

'a'

In [ ]:
data[:1]              #Slicing uses Implicit or Positional indexing

x    a
dtype: object

In [ ]:
data = pd.Series(['a', 'b', 'c'], index=[1, 2, 2])
data

1    a
2    b
2    c
dtype: object

In [ ]:
data[2]

2    b
2    c
dtype: object

* In this example data is same as previous example but indexes are repeating .
* So non - unique indexes works in pandas

In [ ]:
#To avoid confusion Pandas has special indexers - loc and iloc 

''' 
    1. loc - works with explicit indexes/labels.
    2. iloc - works with implicit indexes/positional indexes.

'''

' \n    1. loc - works with explicit indexes/labels.\n    2. iloc - works with implicit indexes/positional indexes.\n\n'

In [ ]:
data = pd.Series(['a', 'b', 'c'], index=[1, 2, 2])


In [ ]:
data

1    a
2    b
2    c
dtype: object

In [ ]:
#loc -

data.loc[1]

'a'

In [ ]:
data.loc[2]

2    b
2    c
dtype: object

In [ ]:
#iloc -

data.iloc[1]

'b'

In [ ]:
data.iloc[2]

'c'

In [ ]:
#Let's come to our dataset and use 'loc' and 'iloc' over the dataset -

df

country  year  population continent  life_exp     gdp_cap
0     Afghanistan  1952     8425333      Asia    28.801  779.445314
1     Afghanistan  1957     9240934      Asia    30.332  820.853030
2     Afghanistan  1962    10267083      Asia    31.997  853.100710
3     Afghanistan  1967    11537966      Asia    34.020  836.197138
4     Afghanistan  1972    13079460      Asia    36.088  739.981106
...           ...   ...         ...       ...       ...         ...
1699     Zimbabwe  1987     9216418    Africa    62.351  706.157306
1700     Zimbabwe  1992    10704340    Africa    60.377  693.420786
1701     Zimbabwe  1997    11404948    Africa    46.809  792.449960
1702     Zimbabwe  2002    11926563    Africa    39.989  672.038623
1703     Zimbabwe  2007    12311143    Africa    43.487  469.709298

[1704 rows x 6 columns]

In [ ]:
#We want to access multiple rows at indexes 1,10,100 using 'iloc' -

df.iloc[[1,10,100]]

country  year  population continent  life_exp     gdp_cap
1    Afghanistan  1957     9240934      Asia    30.332  820.853030
10   Afghanistan  2002    25268405      Asia    42.129  726.734055
100   Bangladesh  1972    70759295      Asia    45.252  630.233627

In [ ]:
#What will happen if we provide same indexes to 'loc' ?? 

df.loc[[1,10,100]]

country  year  population continent  life_exp     gdp_cap
1    Afghanistan  1957     9240934      Asia    30.332  820.853030
10   Afghanistan  2002    25268405      Asia    42.129  726.734055
100   Bangladesh  1972    70759295      Asia    45.252  630.233627

* Output of ' loc ' and ' iloc ' for indexes [ 1 , 10 , 100 ] is same .

* Remember most of the times your indexes would be same so output for ' loc ' and ' iloc ' would be same .

In [ ]:
#What do you think, what will be output of df.iloc[-1] ??

df.iloc[-1]           #This will give last row of dataset .

country         Zimbabwe
year                2007
population      12311143
continent         Africa
life_exp          43.487
gdp_cap       469.709298
Name: 1703, dtype: object

* iloc uses positional indexes which is basically the original natural indexing so ' - 1 ' is valid index for iloc  .

In [ ]:
#Let's check type of this 

type(df.iloc[-1])        

pandas.core.series.Series

* Row is also a series .

In [ ]:
#What will happen if we do df.loc[-1] ??

df.loc[-1]         #It is giving an error

KeyError: ignored

* What do you think why this will give an error ??
* ' - 1 ' is not an explicit index and hence we get an error . invalid index [ -1 ]

In [ ]:
#Let's convert country column into row labels/keys only 

''' To do this we can use set_index() function '''

df

country  year  population continent  life_exp     gdp_cap
0     Afghanistan  1952     8425333      Asia    28.801  779.445314
1     Afghanistan  1957     9240934      Asia    30.332  820.853030
2     Afghanistan  1962    10267083      Asia    31.997  853.100710
3     Afghanistan  1967    11537966      Asia    34.020  836.197138
4     Afghanistan  1972    13079460      Asia    36.088  739.981106
...           ...   ...         ...       ...       ...         ...
1699     Zimbabwe  1987     9216418    Africa    62.351  706.157306
1700     Zimbabwe  1992    10704340    Africa    60.377  693.420786
1701     Zimbabwe  1997    11404948    Africa    46.809  792.449960
1702     Zimbabwe  2002    11926563    Africa    39.989  672.038623
1703     Zimbabwe  2007    12311143    Africa    43.487  469.709298

[1704 rows x 6 columns]

In [ ]:
temp = df.set_index("country")          

In [ ]:
temp

year  population continent  life_exp     gdp_cap
country                                                      
Afghanistan  1952     8425333      Asia    28.801  779.445314
Afghanistan  1957     9240934      Asia    30.332  820.853030
Afghanistan  1962    10267083      Asia    31.997  853.100710
Afghanistan  1967    11537966      Asia    34.020  836.197138
Afghanistan  1972    13079460      Asia    36.088  739.981106
...           ...         ...       ...       ...         ...
Zimbabwe     1987     9216418    Africa    62.351  706.157306
Zimbabwe     1992    10704340    Africa    60.377  693.420786
Zimbabwe     1997    11404948    Africa    46.809  792.449960
Zimbabwe     2002    11926563    Africa    39.989  672.038623
Zimbabwe     2007    12311143    Africa    43.487  469.709298

[1704 rows x 5 columns]

In [ ]:
#Let's extract life_exp column

temp["life_exp"]

country
Afghanistan    28.801
Afghanistan    30.332
Afghanistan    31.997
Afghanistan    34.020
Afghanistan    36.088
                ...  
Zimbabwe       62.351
Zimbabwe       60.377
Zimbabwe       46.809
Zimbabwe       39.989
Zimbabwe       43.487
Name: life_exp, Length: 1704, dtype: float64

* It extracts life_exp column along with there keys/indexes .

In [ ]:
#Let's say we want to extract life_exp of a particular country 

#let's say 'zimbabwe'

temp["life_exp"].loc["zimbabwe"]

KeyError: ignored

In [ ]:
temp["life_exp"].loc["Zimbabwe"]        #Usually recommended to keep row labels unique .

country
Zimbabwe    48.451
Zimbabwe    50.469
Zimbabwe    52.358
Zimbabwe    53.995
Zimbabwe    55.635
Zimbabwe    57.674
Zimbabwe    60.363
Zimbabwe    62.351
Zimbabwe    60.377
Zimbabwe    46.809
Zimbabwe    39.989
Zimbabwe    43.487
Name: life_exp, dtype: float64

* It will extract life_exp for ' Zimbabwe ' .

* This how we can use index to our benefit if we have this kind of special index .

* But it's a good idea to keep the index/labels value for each row unique .

* Usually recommended to keep row labels unique .

In [ ]:
#Let's get back to original dataset 'df'

df

country  year  population continent  life_exp     gdp_cap
0     Afghanistan  1952     8425333      Asia    28.801  779.445314
1     Afghanistan  1957     9240934      Asia    30.332  820.853030
2     Afghanistan  1962    10267083      Asia    31.997  853.100710
3     Afghanistan  1967    11537966      Asia    34.020  836.197138
4     Afghanistan  1972    13079460      Asia    36.088  739.981106
...           ...   ...         ...       ...       ...         ...
1699     Zimbabwe  1987     9216418    Africa    62.351  706.157306
1700     Zimbabwe  1992    10704340    Africa    60.377  693.420786
1701     Zimbabwe  1997    11404948    Africa    46.809  792.449960
1702     Zimbabwe  2002    11926563    Africa    39.989  672.038623
1703     Zimbabwe  2007    12311143    Africa    43.487  469.709298

[1704 rows x 6 columns]

In [ ]:
 #Let's learn how can we add a row 

''' Earlier we have learn that, you can access one of the colum  as a dictionary which means you can use column name as key to access a particular column
    So  one can possibely be thought as a dictionary . 
    
    If you want to add a row then 1st step is to create dictionary '''

d = {'country': 'India', 'year': 2000,'life_exp':37.08,
     'population':135000, "continent": "Asia", "gdp_cap": 100}


In [ ]:
d

{'country': 'India',
 'year': 2000,
 'life_exp': 37.08,
 'population': 135000,
 'continent': 'Asia',
 'gdp_cap': 100}

In [ ]:
#To add this dictionary as a row in data set you can use

#I] Append function

df.append(d)


TypeError: ignored

In [ ]:
df.append(d,ignore_index = True)

country  year  population continent  life_exp     gdp_cap
0     Afghanistan  1952     8425333      Asia    28.801  779.445314
1     Afghanistan  1957     9240934      Asia    30.332  820.853030
2     Afghanistan  1962    10267083      Asia    31.997  853.100710
3     Afghanistan  1967    11537966      Asia    34.020  836.197138
4     Afghanistan  1972    13079460      Asia    36.088  739.981106
...           ...   ...         ...       ...       ...         ...
1700     Zimbabwe  1992    10704340    Africa    60.377  693.420786
1701     Zimbabwe  1997    11404948    Africa    46.809  792.449960
1702     Zimbabwe  2002    11926563    Africa    39.989  672.038623
1703     Zimbabwe  2007    12311143    Africa    43.487  469.709298
1704        India  2000      135000      Asia    37.080  100.000000

[1705 rows x 6 columns]

* You can see row is added at 1704 .

* Append function adds row at the end of dataset .
* In pandas append function has one argument called as ' ignore_index '

* If ignore_index = False it will not add a row , it will through an error .
* If ignore_index = True it will add a row at the end of dataset .


In [ ]:
#Let's check our dataset

df

country  year  population continent  life_exp     gdp_cap
0     Afghanistan  1952     8425333      Asia    28.801  779.445314
1     Afghanistan  1957     9240934      Asia    30.332  820.853030
2     Afghanistan  1962    10267083      Asia    31.997  853.100710
3     Afghanistan  1967    11537966      Asia    34.020  836.197138
4     Afghanistan  1972    13079460      Asia    36.088  739.981106
...           ...   ...         ...       ...       ...         ...
1699     Zimbabwe  1987     9216418    Africa    62.351  706.157306
1700     Zimbabwe  1992    10704340    Africa    60.377  693.420786
1701     Zimbabwe  1997    11404948    Africa    46.809  792.449960
1702     Zimbabwe  2002    11926563    Africa    39.989  672.038623
1703     Zimbabwe  2007    12311143    Africa    43.487  469.709298

[1704 rows x 6 columns]

* As you notice it is not an inplace change . ( It is not a permanent change ) 

* If you want to add this row permanently you can pass argument ' inplace = True ' it will add that row permanently .

In [ ]:
#This was the way to add a new row using dictionary .

'''
    We can add a row using lists .
    The problem with list is we have to make sure that it's in the same order of  my column names in dataset.
    
    list element == order of column names  
'''

df.loc[len(df.index)] = ['India',2000,13500000, 37.08,900.23]


ValueError: ignored

* If we don't follow list element names  = = order of column names you will get this error .

* Here we forget to add  continent . 

In [ ]:
df.loc[len(df.index)] = ['India',2000,13500000,"Asia", 37.08,900.23]


In [ ]:
df

country  year  population continent  life_exp     gdp_cap
0     Afghanistan  1952     8425333      Asia    28.801  779.445314
1     Afghanistan  1957     9240934      Asia    30.332  820.853030
2     Afghanistan  1962    10267083      Asia    31.997  853.100710
3     Afghanistan  1967    11537966      Asia    34.020  836.197138
4     Afghanistan  1972    13079460      Asia    36.088  739.981106
...           ...   ...         ...       ...       ...         ...
1700     Zimbabwe  1992    10704340    Africa    60.377  693.420786
1701     Zimbabwe  1997    11404948    Africa    46.809  792.449960
1702     Zimbabwe  2002    11926563    Africa    39.989  672.038623
1703     Zimbabwe  2007    12311143    Africa    43.487  469.709298
1704        India  2000    13500000      Asia    37.080  900.230000

[1705 rows x 6 columns]

In [ ]:
#iloc will not work in this case

df.iloc[len(df.index)] = ['India',2000,13500000,"Asia", 37.08,900.23]

IndexError: ignored

* iloc will through an error .
* And that error is ' iloc cannot enlarge its target object '

* This new index dosen't exists yet so it is saying iloc cannot enlarge its target object .

* Positional index dosen't exists , because your data was only till 1703 so positional index 1704 dosen't exists , where as loc can create new indexes .


* iloc can only work with the indexes which exists .

In [ ]:
#The recommended way to add rows in data set is use to append function and using dictionary 

In [ ]:
#Let's learn how to drop rows and drop duplicates from dataset .

df

country  year  population continent  life_exp     gdp_cap
0     Afghanistan  1952     8425333      Asia    28.801  779.445314
1     Afghanistan  1957     9240934      Asia    30.332  820.853030
2     Afghanistan  1962    10267083      Asia    31.997  853.100710
3     Afghanistan  1967    11537966      Asia    34.020  836.197138
4     Afghanistan  1972    13079460      Asia    36.088  739.981106
...           ...   ...         ...       ...       ...         ...
1700     Zimbabwe  1992    10704340    Africa    60.377  693.420786
1701     Zimbabwe  1997    11404948    Africa    46.809  792.449960
1702     Zimbabwe  2002    11926563    Africa    39.989  672.038623
1703     Zimbabwe  2007    12311143    Africa    43.487  469.709298
1704        India  2000    13500000      Asia    37.080  900.230000

[1705 rows x 6 columns]

In [ ]:
#Let's find duplicates from dataset 

''' For this you can use dataset_name.duplicated() function '''

df.duplicated()

0       False
1       False
2       False
3       False
4       False
        ...  
1700    False
1701    False
1702    False
1703    False
1704    False
Length: 1705, dtype: bool

* This will give you a binary mask of wether a particular duplicates are present or not .

* But this output is not readable for me .

* This output is not clear .

In [ ]:
#Let's do one think, lets intentionally add some duplicate rows to make this think interesting  or understandable 

df.loc[len(df.index)] = ['India',2000,13500000,"Asia", 37.08,900.23]

In [ ]:
df

country  year  population continent  life_exp     gdp_cap
0     Afghanistan  1952     8425333      Asia    28.801  779.445314
1     Afghanistan  1957     9240934      Asia    30.332  820.853030
2     Afghanistan  1962    10267083      Asia    31.997  853.100710
3     Afghanistan  1967    11537966      Asia    34.020  836.197138
4     Afghanistan  1972    13079460      Asia    36.088  739.981106
...           ...   ...         ...       ...       ...         ...
1701     Zimbabwe  1997    11404948    Africa    46.809  792.449960
1702     Zimbabwe  2002    11926563    Africa    39.989  672.038623
1703     Zimbabwe  2007    12311143    Africa    43.487  469.709298
1704        India  2000    13500000      Asia    37.080  900.230000
1705        India  2000    13500000      Asia    37.080  900.230000

[1706 rows x 6 columns]

* Now we have duplicated rows .

In [ ]:
#Againn lets find duplicates from dataset 

df.duplicated()

0       False
1       False
2       False
3       False
4       False
        ...  
1701    False
1702    False
1703    False
1704    False
1705     True
Length: 1706, dtype: bool

* Here we get one ' True ' it means we have one duplicate row .

* Again this output is not clear and we cannot conclude anything from it .

* We cannot see complete output .

In [ ]:
#So what we can do is, we can use this output (output of duplicated function) as a mask .

df.loc[df.duplicated()]

country  year  population continent  life_exp  gdp_cap
1705   India  2000    13500000      Asia     37.08   900.23

* It will print all duplicate rows from dataset .
* We basically uses a mask to print the rows which are duplicated .

In [ ]:
#Once we have duplicate data in dataset there is one more handy function called as drop_duplicate(), we can use this function to drop duplicates 

'''
drop_duplicate() function has one argument 'keep', this asks you to tell which duplicate you want to drop .
Syntax - drop_duplicate(keep = " ")

This keep can take three values that, are first,last,False

first : first means keep first appearance.
last :  last means keep last appearance.
False : False means delete all appearance.

'''

df.drop_duplicates(keep = "first")

country  year  population continent  life_exp     gdp_cap
0     Afghanistan  1952     8425333      Asia    28.801  779.445314
1     Afghanistan  1957     9240934      Asia    30.332  820.853030
2     Afghanistan  1962    10267083      Asia    31.997  853.100710
3     Afghanistan  1967    11537966      Asia    34.020  836.197138
4     Afghanistan  1972    13079460      Asia    36.088  739.981106
...           ...   ...         ...       ...       ...         ...
1700     Zimbabwe  1992    10704340    Africa    60.377  693.420786
1701     Zimbabwe  1997    11404948    Africa    46.809  792.449960
1702     Zimbabwe  2002    11926563    Africa    39.989  672.038623
1703     Zimbabwe  2007    12311143    Africa    43.487  469.709298
1704        India  2000    13500000      Asia    37.080  900.230000

[1705 rows x 6 columns]

* This has removed the row which was with label 1705 .

In [ ]:
df.drop_duplicates(keep = "last")

country  year  population continent  life_exp     gdp_cap
0     Afghanistan  1952     8425333      Asia    28.801  779.445314
1     Afghanistan  1957     9240934      Asia    30.332  820.853030
2     Afghanistan  1962    10267083      Asia    31.997  853.100710
3     Afghanistan  1967    11537966      Asia    34.020  836.197138
4     Afghanistan  1972    13079460      Asia    36.088  739.981106
...           ...   ...         ...       ...       ...         ...
1700     Zimbabwe  1992    10704340    Africa    60.377  693.420786
1701     Zimbabwe  1997    11404948    Africa    46.809  792.449960
1702     Zimbabwe  2002    11926563    Africa    39.989  672.038623
1703     Zimbabwe  2007    12311143    Africa    43.487  469.709298
1705        India  2000    13500000      Asia    37.080  900.230000

[1705 rows x 6 columns]

* Here you can notice after 1703 we directly have 1705 .
* This has removed the row which was with label 1704 .


* This is not inplace change .

In [ ]:
df

country  year  population continent  life_exp     gdp_cap
0     Afghanistan  1952     8425333      Asia    28.801  779.445314
1     Afghanistan  1957     9240934      Asia    30.332  820.853030
2     Afghanistan  1962    10267083      Asia    31.997  853.100710
3     Afghanistan  1967    11537966      Asia    34.020  836.197138
4     Afghanistan  1972    13079460      Asia    36.088  739.981106
...           ...   ...         ...       ...       ...         ...
1701     Zimbabwe  1997    11404948    Africa    46.809  792.449960
1702     Zimbabwe  2002    11926563    Africa    39.989  672.038623
1703     Zimbabwe  2007    12311143    Africa    43.487  469.709298
1704        India  2000    13500000      Asia    37.080  900.230000
1705        India  2000    13500000      Asia    37.080  900.230000

[1706 rows x 6 columns]

In [ ]:
#Let's learn how to delete rows -

'''
    We want to drop row with label '1' . so we can use drop function to delete a row.
    We have learned drop function works with labels.
    
'''

df.drop(1,axis = 0,inplace = True)

In [ ]:
df

country  year  population continent  life_exp     gdp_cap
0     Afghanistan  1952     8425333      Asia    28.801  779.445314
2     Afghanistan  1962    10267083      Asia    31.997  853.100710
3     Afghanistan  1967    11537966      Asia    34.020  836.197138
4     Afghanistan  1972    13079460      Asia    36.088  739.981106
5     Afghanistan  1977    14880372      Asia    38.438  786.113360
...           ...   ...         ...       ...       ...         ...
1701     Zimbabwe  1997    11404948    Africa    46.809  792.449960
1702     Zimbabwe  2002    11926563    Africa    39.989  672.038623
1703     Zimbabwe  2007    12311143    Africa    43.487  469.709298
1704        India  2000    13500000      Asia    37.080  900.230000
1705        India  2000    13500000      Asia    37.080  900.230000

[1705 rows x 6 columns]

* You can notice row with label ' 1 ' is missing .
* Also notice by default the positional indexes would be changed but row labels haven't change .

* The row labels remains same .

In [ ]:
df.iloc[1]

country       Afghanistan
year                 1962
population       10267083
continent            Asia
life_exp           31.997
gdp_cap         853.10071
Name: 2, dtype: object

* You can notice this ' 1 ' has become the second row .

* iloc has changed itself or positional indexes get changed automatically .

In [ ]:
#Let's see what we will get with loc[1]

df.loc[1]

KeyError: ignored

* This gives an error , because it couldn't find any row with label ' 1 ' .

In [ ]:
#Workung with rows and columns together 

In [ ]:
'''
 Let's say we want to extract rows starting from 2nd row until the 4th row and columns starting from 0 until 3 
    
    How can we do that -

 '''

df.iloc[1:5 , 0:4]

country  year  population continent
2  Afghanistan  1962    10267083      Asia
3  Afghanistan  1967    11537966      Asia
4  Afghanistan  1972    13079460      Asia
5  Afghanistan  1977    14880372      Asia

* Notice that it has given only foure columns because we have mentioned 0 to 4 columns .


* iloc is working with columns with their positional indexes .

In [ ]:
#Let's check dataset


df

country  year  population continent  life_exp     gdp_cap
0     Afghanistan  1952     8425333      Asia    28.801  779.445314
2     Afghanistan  1962    10267083      Asia    31.997  853.100710
3     Afghanistan  1967    11537966      Asia    34.020  836.197138
4     Afghanistan  1972    13079460      Asia    36.088  739.981106
5     Afghanistan  1977    14880372      Asia    38.438  786.113360
...           ...   ...         ...       ...       ...         ...
1701     Zimbabwe  1997    11404948    Africa    46.809  792.449960
1702     Zimbabwe  2002    11926563    Africa    39.989  672.038623
1703     Zimbabwe  2007    12311143    Africa    43.487  469.709298
1704        India  2000    13500000      Asia    37.080  900.230000
1705        India  2000    13500000      Asia    37.080  900.230000

[1705 rows x 6 columns]

In [ ]:
#What do you think if we do loc[1:5 , 0:4] it will work or not 

df.loc[1:5,0:4]

TypeError: ignored

* This will not work because , clearly we can see ' 1 ' which is not present so this coulde be one of the reason but this is actually not a reason .

* So reason is loc work with labels and labels for the column are column names not this positional indexes .


* We have to pass list of column names which we need , loc works with labels .

* So we are getting an error .

In [ ]:
#Let's try using some labels

df.loc[1:5 ,["country" , "life_exp"]]          #This will work even if 1 isn't there

country  life_exp
2  Afghanistan    31.997
3  Afghanistan    34.020
4  Afghanistan    36.088
5  Afghanistan    38.438

* Although ' 1 ' is present it will work for us .

* It started with  ' 2 ' and neglected  ' 1 ' .

* Notice one more thing in above output is  -

* ' 5 ' is also included ,  Which means  ' slicing ' with loc considers the end point as well .

In [ ]:
df.loc[1:5 , "year":"population"]        #This also works for us

year  population
2  1962    10267083
3  1967    11537966
4  1972    13079460
5  1977    14880372

In [ ]:
df.loc[1:5, "year":"life_exp"]

year  population continent  life_exp
2  1962    10267083      Asia    31.997
3  1967    11537966      Asia    34.020
4  1972    13079460      Asia    36.088
5  1977    14880372      Asia    38.438

* Notice that this kind of slicing work for loc . and end point is also inclusive .

* Note - loc always will works with labels which are mentioned here . ( column names , row labels like  2 , 3 , 4 , 5   ..... )


In [ ]:
#We can access multiple rows and columns 

''' Here we are accessing rows with positional indexes 0,10,100 and columns with positional indexes 0,2,3 '''

df.iloc[[0,10,100],[0,2,3]]

country  population continent
0    Afghanistan     8425333      Asia
11   Afghanistan    31889923      Asia
101   Bangladesh    80428306      Asia

In [ ]:
df.iloc[1:10:2]

country  year  population continent  life_exp     gdp_cap
2   Afghanistan  1962    10267083      Asia    31.997  853.100710
4   Afghanistan  1972    13079460      Asia    36.088  739.981106
6   Afghanistan  1982    12881816      Asia    39.854  978.011439
8   Afghanistan  1992    16317921      Asia    41.674  649.341395
10  Afghanistan  2002    25268405      Asia    42.129  726.734055

In [ ]:
#Basic operations in pandas

''' Let's access life_exp column '''

df["life_exp"]

0       28.801
2       31.997
3       34.020
4       36.088
5       38.438
         ...  
1701    46.809
1702    39.989
1703    43.487
1704    37.080
1705    37.080
Name: life_exp, Length: 1705, dtype: float64

In [ ]:
#We want to find mean of life_exp column

df["life_exp"].mean()

59.46526256891495

In [ ]:
#We want to find sum of life_exp column 

df["life_exp"].sum()

101388.27268

In [ ]:
#Let's say we want to find count

df["life_exp"].count()

1705

In [ ]:
#Now let's check output of above operations are correct or not 

''' For that you can do this - sum divided by count is equal to mean '''

df["life_exp"].sum() / df["life_exp"].count()

59.46526256891495

* This is equal to mean so above all outputs are correct .

In [ ]:
#Sorting in pandas 

'''
 Our dataset is not sorted and we may want to sort it as per somre particular way .
 lets say we want to sort the rows as per the year.
 
 For this you can use sort_values() function

'''

df.sort_values(["year"])

country  year  population continent  life_exp  \
0                Afghanistan  1952     8425333      Asia    28.801   
924                   Malawi  1952     2917802    Africa    36.256   
144   Bosnia and Herzegovina  1952     2791000    Europe    53.820   
912               Madagascar  1952     4762912    Africa    36.681   
900                    Libya  1952     1019729    Africa    42.723   
...                      ...   ...         ...       ...       ...   
1211                    Peru  2007    28674757  Americas    71.421   
1223             Philippines  2007    91077287      Asia    71.688   
1235                  Poland  2007    38518241    Europe    75.563   
143                  Bolivia  2007     9119152  Americas    65.554   
131                    Benin  2007     8078314    Africa    56.728   

           gdp_cap  
0       779.445314  
924     369.165080  
144     973.533195  
912    1443.011715  
900    2387.548060  
...            ...  
1211   7408.905561  
1223   3190.481016  
1235  15389.924680  
143    3822.137084  
131    1441.284873  

[1705 rows x 6 columns]

* Sorted by year .

In [ ]:
#Let's say we are insterested in sorting in descending order -

'''
    So sort_values() function has argument called as ascending .
    By default value of ascending is True 

'''

df.sort_values(["year"],ascending = False)     #Sorts a dataset in descending order

country  year  population continent  life_exp  \
143                  Bolivia  2007     9119152  Americas    65.554   
1451                   Sudan  2007    42292929    Africa    58.556   
443       Dominican Republic  2007     9319622  Americas    72.235   
155   Bosnia and Herzegovina  2007     4552198    Europe    74.852   
803                    Japan  2007   127467972      Asia    82.603   
...                      ...   ...         ...       ...       ...   
624            Guinea-Bissau  1952      580653    Africa    32.500   
612                   Guinea  1952     2664249    Africa    33.609   
1596          United Kingdom  1952    50430000    Europe    69.180   
1164                Pakistan  1952    41346560      Asia    43.436   
0                Afghanistan  1952     8425333      Asia    28.801   

           gdp_cap  
143    3822.137084  
1451   2602.394995  
443    6025.374752  
155    7446.298803  
803   31656.068060  
...            ...  
624     299.850319  
612     510.196492  
1596   9979.508487  
1164    684.597144  
0       779.445314  

[1705 rows x 6 columns]

In [ ]:
#Sort using life_exp column

df.sort_values(["life_exp"])

country  year  population continent  life_exp       gdp_cap
1292            Rwanda  1992     7290203    Africa    23.599    737.068595
0          Afghanistan  1952     8425333      Asia    28.801    779.445314
552             Gambia  1952      284320    Africa    30.000    485.230659
36              Angola  1952     4232095    Africa    30.015   3520.610273
1344      Sierra Leone  1952     2143249    Africa    30.331    879.787736
...                ...   ...         ...       ...       ...           ...
1487       Switzerland  2007     7554661    Europe    81.701  37506.419070
695            Iceland  2007      301931    Europe    81.757  36180.789190
802              Japan  2002   127065841      Asia    82.000  28604.591900
671   Hong Kong, China  2007     6980412      Asia    82.208  39724.978670
803              Japan  2007   127467972      Asia    82.603  31656.068060

[1705 rows x 6 columns]

* Sorted by life_exp .

In [ ]:
#Let's say we have to sort by both year and life_exp columns

df.sort_values(["year", "life_exp"])

country  year  population continent  life_exp       gdp_cap
0          Afghanistan  1952     8425333      Asia    28.801    779.445314
552             Gambia  1952      284320    Africa    30.000    485.230659
36              Angola  1952     4232095    Africa    30.015   3520.610273
1344      Sierra Leone  1952     2143249    Africa    30.331    879.787736
1032        Mozambique  1952     6446316    Africa    31.286    468.526038
...                ...   ...         ...       ...       ...           ...
71           Australia  2007    20434176   Oceania    81.235  34435.367440
1487       Switzerland  2007     7554661    Europe    81.701  37506.419070
695            Iceland  2007      301931    Europe    81.757  36180.789190
671   Hong Kong, China  2007     6980412      Asia    82.208  39724.978670
803              Japan  2007   127467972      Asia    82.603  31656.068060

[1705 rows x 6 columns]

In [ ]:
#Sort 'year' in descending order and 'life_exp' in ascending order

df.sort_values(["year", "life_exp"], ascending=[False, True])


country  year  population continent  life_exp       gdp_cap
1463     Swaziland  2007     1133066    Africa    39.613   4513.480643
1043    Mozambique  2007    19951656    Africa    42.082    823.685621
1691        Zambia  2007    11746035    Africa    42.384   1271.211593
1355  Sierra Leone  2007     6144562    Africa    42.568    862.540756
887        Lesotho  2007     2012649    Africa    42.592   1569.331442
...            ...   ...         ...       ...       ...           ...
408        Denmark  1952     4334000    Europe    70.780   9692.385245
1464        Sweden  1952     7124673    Europe    71.860   8527.844662
1080   Netherlands  1952    10381988    Europe    72.130   8941.571858
684        Iceland  1952      147962    Europe    72.490   7267.688428
1140        Norway  1952     3327728    Europe    72.670  10095.421720

[1705 rows x 6 columns]

* In ascending  = True argument  we can pass list of True and False .

In [ ]:
#Let's learn how to create a dataframe from scratch

df.loc[0]

country       Afghanistan
year                 1952
population        8425333
continent            Asia
life_exp           28.801
gdp_cap        779.445314
Name: 0, dtype: object

* If you extract a row it's a series .

In [ ]:
df["country"]

0       Afghanistan
2       Afghanistan
3       Afghanistan
4       Afghanistan
5       Afghanistan
           ...     
1701       Zimbabwe
1702       Zimbabwe
1703       Zimbabwe
1704          India
1705          India
Name: country, Length: 1705, dtype: object

* If you extract a column that's also serise .

In [ ]:
#Let's create series from scratch

pd.Series([10, 20, 30], index=["a", "b", "c"])

a    10
b    20
c    30
dtype: int64

In [ ]:
#Let's create dataframe 

'''
There are two approaches to create dataframe

I] Approach - 1 : Row-Oriented
II]Approch - 2 : Column Oriented

'''

In [ ]:
#I] Approach - 1 : Row-Oriented


pd.DataFrame([[10, 20], [30, 40]], columns=["A", "B"])


A   B
0  10  20
1  30  40

* In Row-Oriented Approach we pass a data as list of list .

* By default it takes row labels  ' 0 '  and   ' 1 ' .

In [ ]:
#II]Approch - 2 : Column Oriented


pd.DataFrame({"A": [10, 30], "B": [20, 40]})

A   B
0  10  20
1  30  40

* In Column-Oriented Approach we pass a data as dictionary .

* By default it takes row labels  ' 0 '  and   ' 1 ' .

In [ ]:
#Pending: Concatenation of dataframes, merging/joining of dataframe - please ignore
#Functions used are - concat(), merge()

In [ ]:
## Assignments -

In [ ]:
#2 - Given the following data frame "df", which of the following command(s) is/are the correct way to extract the mentioned columns in the order: time, total_bill, tip ?

#Option A,B,C are correct answers

In [ ]:


import pandas as pd
import numpy as np 

In [ ]:
#Dataset -

!gdown https://d2beiqkhq929f0.cloudfront.net/public_assets/assets/000/002/276/original/tips.csv?1645193273

Downloading...
From: https://d2beiqkhq929f0.cloudfront.net/public_assets/assets/000/002/276/original/tips.csv?1645193273
To: /content/tips.csv?1645193273
100% 7.94k/7.94k [00:00<00:00, 26.0MB/s]


In [ ]:
pd.read_csv("tips.csv?1645193273") 

total_bill   tip     sex smoker   day    time  size
0         16.99  1.01  Female     No   Sun  Dinner     2
1         10.34  1.66    Male     No   Sun  Dinner     3
2         21.01  3.50    Male     No   Sun  Dinner     3
3         23.68  3.31    Male     No   Sun  Dinner     2
4         24.59  3.61  Female     No   Sun  Dinner     4
..          ...   ...     ...    ...   ...     ...   ...
239       29.03  5.92    Male     No   Sat  Dinner     3
240       27.18  2.00  Female    Yes   Sat  Dinner     2
241       22.67  2.00    Male    Yes   Sat  Dinner     2
242       17.82  1.75    Male     No   Sat  Dinner     2
243       18.78  3.00  Female     No  Thur  Dinner     2

[244 rows x 7 columns]

In [ ]:
df = pd.read_csv("tips.csv?1645193273") 

In [ ]:
df

total_bill   tip     sex smoker   day    time  size
0         16.99  1.01  Female     No   Sun  Dinner     2
1         10.34  1.66    Male     No   Sun  Dinner     3
2         21.01  3.50    Male     No   Sun  Dinner     3
3         23.68  3.31    Male     No   Sun  Dinner     2
4         24.59  3.61  Female     No   Sun  Dinner     4
..          ...   ...     ...    ...   ...     ...   ...
239       29.03  5.92    Male     No   Sat  Dinner     3
240       27.18  2.00  Female    Yes   Sat  Dinner     2
241       22.67  2.00    Male    Yes   Sat  Dinner     2
242       17.82  1.75    Male     No   Sat  Dinner     2
243       18.78  3.00  Female     No  Thur  Dinner     2

[244 rows x 7 columns]

In [ ]:
df.head()

total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3
3       23.68  3.31    Male     No  Sun  Dinner     2
4       24.59  3.61  Female     No  Sun  Dinner     4

In [ ]:
#extract the mentioned columns in the order: time, total_bill, tip ?


#Option A

pd.DataFrame(df, columns=['time', 'total_bill', 'tip'])

time  total_bill   tip
0    Dinner       16.99  1.01
1    Dinner       10.34  1.66
2    Dinner       21.01  3.50
3    Dinner       23.68  3.31
4    Dinner       24.59  3.61
..      ...         ...   ...
239  Dinner       29.03  5.92
240  Dinner       27.18  2.00
241  Dinner       22.67  2.00
242  Dinner       17.82  1.75
243  Dinner       18.78  3.00

[244 rows x 3 columns]

In [ ]:


#Option B

df[['time', 'total_bill', 'tip']]

time  total_bill   tip
0    Dinner       16.99  1.01
1    Dinner       10.34  1.66
2    Dinner       21.01  3.50
3    Dinner       23.68  3.31
4    Dinner       24.59  3.61
..      ...         ...   ...
239  Dinner       29.03  5.92
240  Dinner       27.18  2.00
241  Dinner       22.67  2.00
242  Dinner       17.82  1.75
243  Dinner       18.78  3.00

[244 rows x 3 columns]

In [ ]:
#Option C

df.loc[:, ['time', 'total_bill', 'tip']]

time  total_bill   tip
0    Dinner       16.99  1.01
1    Dinner       10.34  1.66
2    Dinner       21.01  3.50
3    Dinner       23.68  3.31
4    Dinner       24.59  3.61
..      ...         ...   ...
239  Dinner       29.03  5.92
240  Dinner       27.18  2.00
241  Dinner       22.67  2.00
242  Dinner       17.82  1.75
243  Dinner       18.78  3.00

[244 rows x 3 columns]

In [ ]:
#Option D

df.iloc[:,0:2]

total_bill   tip
0         16.99  1.01
1         10.34  1.66
2         21.01  3.50
3         23.68  3.31
4         24.59  3.61
..          ...   ...
239       29.03  5.92
240       27.18  2.00
241       22.67  2.00
242       17.82  1.75
243       18.78  3.00

[244 rows x 2 columns]

In [ ]:
df.iloc[:2,:2]

total_bill   tip
0       16.99  1.01
1       10.34  1.66